packages/app-builder-lib/src/node-module-collector/npmNodeModulesCollector.ts

```diff
@@ -15,22 +15,39 @@ export class NpmNodeModulesCollector extends NodeModulesCollector {
     return ["list", "-a", "--include", "prod", "--include", "optional", "--omit", "dev", "--json", "--long", "--silent"]
   }
 
-  removeNonProductionDependencie(tree: DependencyTree) {
-    const dependencies = tree.dependencies || {}
-    const _dependencies = tree._dependencies || {}
-    if (Object.keys(_dependencies).length > 0 && Object.keys(dependencies).length === 0) {
-      tree.dependencies = this.allDependencies.get(`${tree.name}@${tree.version}`)?.dependencies || {}
-      tree.__circularDependencyDetected = true
-      log.debug({ name: tree.name, version: tree.version }, "circular dependency detected")
-      return
+  protected removeNonProductionDependencies(tree: DependencyTree): DependencyTree {
+    const _deps = tree._dependencies ?? {}
+    const deps = tree.dependencies ?? {}
+    if (Object.keys(_deps).length > 0 && Object.keys(deps).length === 0) {
+      tree.dependencies = this.allDependencies.get(`${tree.name}@${tree.version}`)?.dependencies
+      tree.implicitDependenciesInjected = true
+      log.debug({ name: tree.name, version: tree.version }, "injecting implicit _dependencies")
+      return tree
     }
 
-    for (const [key, value] of Object.entries(dependencies)) {
-      if (!_dependencies[key] || Object.keys(value).length === 0) {
-        delete dependencies[key]
-        continue
+    const dependencies = Object.entries(tree.dependencies || {}).reduce<DependencyTree["dependencies"]>((acc, curr) => {
```

Comment:
Will the performance of the `reduce` function be better? Currently, the `reduce` syntax seems a bit complex and not as simple or readable. Using a `for` loop appears to be more straightforward.

Reply:
From what I have read, `reduce` is more performant. Old ref: https://dev.to/babak/why-you-should-use-reduce-instead-of-loops----part-i-5dfa

It also creates a new data struct, which I am deliberately doing instead of manipulating an object outside of the loop. The `delete` logic was manipulating the original `JSON.parse` tree, continually, through different logic flows across multiple files. It was making it far harder for me to track and identify the state/changes of data while at breakpoints. With this approach, we are always generating a unique tree. Albeit, yes, memory usage is temporarily increased, but we're working with JSON here and yarn/npm process just fine.